tests-ng/plugins/sysdiff.py

```diff
@@ -0,0 +1,482 @@
+"""
+System state snapshot and diff functionality for tests-ng.
+
+Provides snapshot and comparison capabilities for:
+- packages (dpkg)
+- systemd units
+- files (sha256 checksums)
+- sysctl parameters
+- loaded kernel modules
+"""
+
+import difflib
+import json
+import os
+import re
+import tempfile
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import pytest
+
+from .shell import ShellRunner
+from .dpkg import Dpkg
+from .systemd import Systemd, SystemdUnit
+from .kernel_module import KernelModule
+from .sysctl import Sysctl
+
+
+DEFAULT_PATHS = [
+    "/etc", "/boot", "/usr/local/bin", "/usr/local/sbin",
+    "/usr/local/lib", "/opt", "/proc/mounts"
+]
+
+STATE_DIR = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local/state")) / "sysdiff"
```

Comment:
the sysdiff test can only be run as root, thus this home directory is always root's home directory.. not sure if that is the right place to put this

also, [the spec](https://specifications.freedesktop.org/basedir-spec/latest/#variables) mentions that this dir is supposed for state that is available for multiple runs of the application, but we are deleting it after the test has run

not exactly sure what the ideal solution would be and I think the way it is is good enough, but maybe there is a more fitting place for this

Reply:
sysdiff can be run as any user but having root access is needed to get the full picture (e.g. file access).
Initially I planned to have a `sysdiff` cli command and that was the reason to put the output to the user's $HOME.
I could also put the outputs to `/tmp/` to be gone the next reboot :shrug: and do not have a strong opinion here.

I'll change it in a way that `~/.local/state/sysdiff/${DATE}-${NAME}.json` is now the format to save snapshots. This looks more like "multiple runs" are supported.
Also, I'll add back the CLI part so that there is a way to easily test changes to `tests-ng/plugins/sysdiff.py` by calling `tests-ng/util/sysdiff-cli.py`.